lib/cc/service/safe_webhook.rb

```diff
@@ -0,0 +1,54 @@
+module CC
+  class Service
+    class SafeWebhook
+      InvalidWebhookURL = Class.new(StandardError)
+
+      def self.getaddress(host)
+        @resolv ||= Resolv::DNS.new
+        @resolv.getaddress(host)
+      end
+
+      def initialize(url)
+        @url = url
+      end
+
+      # Resolve the Host to an IP address, validate that it doesn't point to
+      # anything internal, then alter the request to be for the IP directly with
+      # an explicit Host header given.
+      #
+      # See http://blog.fanout.io/2014/01/27/how-to-safely-invoke-webhooks/#ip-address-blacklisting
+      def validate!(request)
+        uri = URI.parse(url)
+        address = self.class.getaddress(uri.host)
+
+        if internal?(address)
+          raise_invalid("resolves to a private IP address")
+        end
+
+        alter_request(request, uri, address)
+      rescue URI::InvalidURIError, Resolv::ResolvError => ex
+        raise_invalid(ex.message)
+      end
+
+      private
+
+      attr_reader :url
+
+      def internal?(address)
+        address.start_with?("127.") ||
+          address.start_with?("10.")
+      end
+
+      def alter_request(request, uri, address)
+        address_uri = uri.dup
+        address_uri.host = address
+        request.url(address_uri.to_s)
+        request.headers.update(Host: uri.host)
+      end
+
+      def raise_invalid(message)
+        raise InvalidWebhookURL, "The Webhook URL #{url} is invalid: #{message}"
```

Comment:
I'm a bit surprised we're just raising this, and not using the catch above to add a validation error of some kind. Are you planning to catch this app-side & do something with it or just let it blow up?